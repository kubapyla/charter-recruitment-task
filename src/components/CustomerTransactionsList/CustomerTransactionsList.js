import React, { useState } from 'react';
import { useParams } from 'react-router-dom';
import {
  Button,
  Heading,
  Stack,
  Stat,
  StatLabel,
  StatNumber,
  Table,
  Tbody,
  Td,
  Th,
  Thead,
  Tr,
} from '@chakra-ui/react';
import { arrayOf, number, shape, string } from 'prop-types';
import {
  getCustomerTransactions,
  getRewardPoints,
  getTransactionsFromMonth,
} from 'utils';
import dayjs from 'dayjs';

const CustomerTransactionsList = ({ customers, transactions }) => {
  const { id } = useParams();
  const [selectedMonth, selectMonth] = useState('ALL');

  const customerInformations = customers.find(
    (customer) => customer.customerId === Number(id)
  );

  const customerTransactions = getCustomerTransactions(transactions, id);

  const renderHeader = () => {
    return (
      <Thead>
        <Tr>
          <Th>Transaction ID</Th>
          <Th>Product name</Th>
          <Th>Transaction date</Th>
          <Th>Amount</Th>
        </Tr>
      </Thead>
    );
  };

  const renderRows = (customerTransactionsList) =>
    customerTransactionsList
      .sort((a, b) =>
        dayjs(b.transactionDate).isAfter(dayjs(a.transactionDate)) ? 1 : -1
      )
      .map((customerTransaction) => (
        <Tr key={customerTransaction.transactionId}>
          <Td>{customerTransaction.transactionId}</Td>
          <Td>{customerTransaction.productName}</Td>
          <Td>{customerTransaction.transactionDate}</Td>
          <Td>{`${customerTransaction.amount} ${customerTransaction.currency}`}</Td>
        </Tr>
      ));

  const renderFilterButton = (monthsBack) => (
    <Button
      onClick={() => selectMonth(monthsBack)}
      disabled={Number(selectedMonth) === monthsBack}
    >
      {dayjs().subtract(monthsBack, 'month').format('MMMM')}
    </Button>
  );

  const renderFilters = () => (
    <>
      <Button
        onClick={() => selectMonth('ALL')}
        disabled={selectedMonth === 'ALL'}
      >
        All transactions
      </Button>
      {renderFilterButton(0)}
      {renderFilterButton(1)}
      {renderFilterButton(2)}
      {renderFilterButton(3)}
    </>
  );

  return (
    <Stack spacing={3} paddingTop={10}>
      <Stack
        direction={'column'}
        spacing={3}
        align={'center'}
        alignSelf={'center'}
        position={'relative'}
      >
        <Stack
          paddingBottom={10}
          minW={500}
          align={'center'}
          alignSelf={'center'}
        >
          <Heading fontWeight={600} lineHeight={'110%'}>
            {customerInformations?.customerName}
          </Heading>
          <Stack direction="row" align={'center'} alignSelf={'center'}>
            <Stat minW={200} align={'center'} alignSelf={'center'}>
              <StatLabel>Total points</StatLabel>
              <StatNumber>{getRewardPoints(customerTransactions)}</StatNumber>
            </Stat>
            {selectedMonth !== 'ALL' && (
              <Stat minW={200} align={'center'} alignSelf={'center'}>
                <StatLabel>Active month points</StatLabel>
                <StatNumber>
                  {getRewardPoints(
                    getTransactionsFromMonth(
                      customerTransactions,
                      selectedMonth
                    )
                  )}
                </StatNumber>
              </Stat>
            )}
          </Stack>
        </Stack>
        <Stack direction={'row'}>{renderFilters()}</Stack>
        <Table variant="simple">
          {renderHeader()}
          <Tbody>
            {renderRows(
              selectedMonth !== 'ALL'
                ? getTransactionsFromMonth(customerTransactions, selectedMonth)
                : customerTransactions
            )}
          </Tbody>
        </Table>
      </Stack>
    </Stack>
  );
};

CustomerTransactionsList.propTypes = {
  customers: arrayOf(shape({ customerId: number, customerName: string })),
  transactions: arrayOf(
    shape({
      amount: number,
      currency: string,
      customer: shape({ customerId: number, customerName: string }),
      productName: string,
      transactionDate: string,
      transactionId: number,
    })
  ),
};

CustomerTransactionsList.defaultProps = {
  customers: [],
  transactions: [],
};

export default CustomerTransactionsList;
