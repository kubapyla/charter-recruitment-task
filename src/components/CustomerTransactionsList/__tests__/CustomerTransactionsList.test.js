import React from 'react';
import { shallow } from 'enzyme';
import CustomerTransactionsList from '../CustomerTransactionsList';
import Router from 'react-router-dom';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
}));

const mockCustomers = [
  { customerId: 1, customerName: 'John Smith' },
  { customerId: 2, customerName: 'Jenny Smith' },
];
const mockCustomerTransactionsList = [
  {
    amount: 249,
    currency: 'USD',
    customer: { customerId: 1, customerName: 'John Smith' },
    productName: 'Table',
    transactionDate: '2021-07-10',
    transactionId: 0,
  },
  {
    amount: 80,
    currency: 'USD',
    customer: { customerId: 1, customerName: 'John Smith' },
    productName: 'Chair',
    transactionDate: '2021-06-10',
    transactionId: 1,
  },
  {
    amount: 394,
    currency: 'USD',
    customer: { customerId: 2, customerName: 'Jenny Smith' },
    productName: 'Lamp',
    transactionDate: '2021-07-11',
    transactionId: 2,
  },
];

describe('CustomerTransactionsList', () => {
  let wrapper;

  it('should match snapshot', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ id: '1' });

    wrapper = shallow(
      <CustomerTransactionsList
        customers={mockCustomers}
        transactions={mockCustomerTransactionsList}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
