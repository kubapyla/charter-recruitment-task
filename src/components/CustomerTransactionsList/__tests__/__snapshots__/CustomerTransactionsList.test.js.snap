// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerTransactionsList should match snapshot 1`] = `
<Stack
  paddingTop={10}
  spacing={3}
>
  <Stack
    align="center"
    alignSelf="center"
    direction="column"
    position="relative"
    spacing={3}
  >
    <Stack
      align="center"
      alignSelf="center"
      minW={500}
      paddingBottom={10}
    >
      <Heading
        fontWeight={600}
        lineHeight="110%"
      >
        John Smith
      </Heading>
      <Stack
        align="center"
        alignSelf="center"
        direction="row"
      >
        <Stat
          align="center"
          alignSelf="center"
          minW={200}
        >
          <StatLabel>
            Total points
          </StatLabel>
          <StatNumber>
            378
          </StatNumber>
        </Stat>
      </Stack>
    </Stack>
    <Stack
      direction="row"
    >
      <Button
        disabled={true}
        onClick={[Function]}
      >
        All transactions
      </Button>
      <Button
        disabled={false}
        onClick={[Function]}
      >
        July
      </Button>
      <Button
        disabled={false}
        onClick={[Function]}
      >
        June
      </Button>
      <Button
        disabled={false}
        onClick={[Function]}
      >
        May
      </Button>
      <Button
        disabled={false}
        onClick={[Function]}
      >
        April
      </Button>
    </Stack>
    <Table
      variant="simple"
    >
      <ForwardRef>
        <ForwardRef>
          <ForwardRef>
            Transaction ID
          </ForwardRef>
          <ForwardRef>
            Product name
          </ForwardRef>
          <ForwardRef>
            Transaction date
          </ForwardRef>
          <ForwardRef>
            Amount
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
      <ForwardRef>
        <ForwardRef
          key="0"
        >
          <ForwardRef>
            0
          </ForwardRef>
          <ForwardRef>
            Table
          </ForwardRef>
          <ForwardRef>
            2021-07-10
          </ForwardRef>
          <ForwardRef>
            249 USD
          </ForwardRef>
        </ForwardRef>
        <ForwardRef
          key="1"
        >
          <ForwardRef>
            1
          </ForwardRef>
          <ForwardRef>
            Chair
          </ForwardRef>
          <ForwardRef>
            2021-06-10
          </ForwardRef>
          <ForwardRef>
            80 USD
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
    </Table>
  </Stack>
</Stack>
`;
